### PR TITLE
#445: PFCOUNT benchmark and caching implementation analysis

### PR DESCRIPTION
**Pull request by Dev79844**
Issue: #445

### Summary:
Added benchmarking for PFCOUNT command

Screenshot 2024-09-15 at 12 39 42 PM

If we look at the above metrics, I guess we should go for caching the results.

### Changes:
- Added BenchmarkEvalPFCOUNT in eval_test.go

### CODE REVIEW

File: internal/eval/eval_test.go [hunk 1]
```diff
@@ -217,39 +215,11 @@ func testEvalEXPIRE(t *testing.T, store *dstore.Store) {
...
-				obj := &object.Obj{
-					Value:          value,
-					LastAccessedAt: uint32(time.Now().Unix()),
-				}
-				store.Put(key, obj)
-
-			},
-			input:  []string{"EXISTING_KEY", strconv.FormatInt(-1, 10)},
-			output: []byte("-ERR invalid expire time in 'expire' command\r\n"),
-		},
```
Comment (JyotinderSingh) [inline internal/eval/eval_test.go:252]: Why did we remove this?

File: internal/eval/eval_test.go [hunk 2]
```diff
@@ -325,38 +296,11 @@ func testEvalEXPIREAT(t *testing.T, store *dstore.Store) {
...
-				obj := &object.Obj{
-					Value:          value,
-					LastAccessedAt: uint32(time.Now().Unix()),
-				}
-				store.Put(key, obj)
-
-			},
-			input:  []string{"EXISTING_KEY", strconv.FormatInt(-1, 10)},
-			output: []byte("-ERR invalid expire time in 'expireat' command\r\n"),
-		},
```
Comment (JyotinderSingh) [inline internal/eval/eval_test.go:359]: Why did we remove this?

File: internal/eval/eval_test.go [hunk 3]
```diff
@@ -382,6 +326,7 @@ func testEvalJSONARRLEN(t *testing.T, store *dstore.Store) {
 				_ = sonic.Unmarshal([]byte(value), &rootData)
 				obj := store.NewObj(rootData, -1, object.ObjTypeJSON, object.ObjEncodingJSON)
 				store.Put(key, obj)
+
```
Comment (JyotinderSingh) [inline internal/eval/eval_test.go:329]: nit: unnecessary new line, same for other occurrences below this.

File: internal/eval/eval_test.go [hunk 4]
```diff
@@ -1507,6 +1421,70 @@ func testEvalHSET(t *testing.T, store *dstore.Store) {
...
+        obj := &object.Obj{
+            Value:          hll,
+            LastAccessedAt: uint32(time.Now().Unix()),
+        }
+        store.Put(key, obj)
+    }
+
+    // Create small HLLs (2 items each)
+    createAndInsertHLL("SMALL1", []string{"Item1", "Item2"})
+    createAndInsertHLL("SMALL2", []string{"Item3", "Item4"})
```
Comment (JyotinderSingh) [inline internal/eval/eval_test.go:1442]: We don't need to benchmark for very small values

File: internal/eval/eval_test.go [hunk 5]
```diff
@@ -1507,6 +1421,70 @@ func testEvalHSET(t *testing.T, store *dstore.Store) {
...
+    for i := 0; i < 100; i++ {
+        mediumItems[i] = fmt.Sprintf("MediumItem%d", i)
+    }
+    createAndInsertHLL("MEDIUM1", mediumItems)
+    createAndInsertHLL("MEDIUM2", mediumItems)
+
+    // Create large HLLs (10000 items each)
+    largeItems := make([]string, 10000)
+    for i := 0; i < 10000; i++ {
+        largeItems[i] = fmt.Sprintf("LargeItem%d", i)
```
Comment (JyotinderSingh) [inline internal/eval/eval_test.go:1455]: It would be better to benchmark for 10^4, 10^6, 10^9 values.

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]
Thanks for adding the benchmarking logic for HLLs in Dice. Have left a few comments.

File: internal/eval/eval_test.go (hunk 1 above)
Comment (AshwinKul28) [reply internal/eval/eval_test.go:252]: These are the new cases added, @Dev79844 Please rebase it with master so this code won't get removed.

File: internal/eval/eval_test.go (hunk 2 above)
Comment (AshwinKul28) [reply internal/eval/eval_test.go:359]: These are the new cases added, @Dev79844 Please rebase it with master so this code won't get removed.

File: internal/eval/eval_test.go (hunk 1 above)
Comment (Dev79844) [reply internal/eval/eval_test.go:253]: I tried to rebase the code but still it isn't showing so how about I add them manually?

Comment (JyotinderSingh) [reply internal/eval/eval_test.go]: I've fixed it by merging the master.

File: internal/eval/eval_test.go (hunk 2 above)
Comment (JyotinderSingh) [reply internal/eval/eval_test.go]: I've fixed it by merging the master.

File: internal/eval/eval_test.go (hunk 3 above)
Comment (JyotinderSingh) [reply internal/eval/eval_test.go]: I've fixed it by merging the master.

**Pull request merged**